README.md
# Microsoft Azure Mailer for Laravel

✅ Simple implementation example of [Symfony Azure Mailer Bridge](https://github.com/symfony/azure-mailer) for Laravel Framework.

✅ Bootable scripts for Laravel 9+

[![Latest Stable Version](http://poser.pugx.org/hafael/azure-mailer-driver/v)](https://packagist.org/packages/hafael/azure-mailer-driver)
[![Latest Unstable Version](http://poser.pugx.org/hafael/azure-mailer-driver/v/unstable)](https://packagist.org/packages/hafael/azure-mailer-driver)
[![Total Downloads](http://poser.pugx.org/hafael/azure-mailer-driver/downloads)](https://packagist.org/packages/hafael/azure-mailer-driver)
[![License](http://poser.pugx.org/hafael/azure-mailer-driver/license)](https://packagist.org/packages/hafael/azure-mailer-driver)

A use case of the [symfony/azure-mailer](https://github.com/symfony/azure-mailer) component using bootable scripts in the Laravel framework to send email messages.


## 💡 Requirements

- PHP 8.2 or higher
- [symfony/azure-mailer](https://github.com/symfony/azure-mailer)


## 🧩 Available resources

| Resource             | Status   |
| -------------------- | :------: |
| Plain Text           | ✅  |
| HTML                 | ✅  |
| Attachments          | ✅  |
| Multiple recipients  | ✅  |
| Auth HMAC-SHA256     | ✅  |
| Notifications        | ✅  |
| Mkt Campaigns        | ✅  |
| Markdown             | ✅  |

## 📦 Installation 

First time using Azure ECS? Create your [Azure account](https://azure.com), if you don’t have one already.

1. Download [Composer](https://getcomposer.org/doc/00-intro.md) if not already installed

2. On your project directory run on the command line
`composer require hafael/azure-mailer-driver`

3. Get your Azure CS Access Key and Service Endpoint.


## 🌟 Set mail config
  
Add entry to [root-of-laravel]/config/mail.php:
  
```php
  <?php
    
    ...

    'mailers' => [
        //...other drivers

        'azure' => [
            'transport'             => 'azure',
            'resource_name'         => env('AZURE_MAIL_RESOURCE_NAME'),
            'endpoint'              => env('AZURE_MAIL_ENDPOINT', 'https://my-acs-resource-name.communication.azure.com'),
            'access_key'            => env('AZURE_MAIL_KEY'),
            'api_version'           => env('AZURE_MAIL_API_VERSION', '2023-03-31'),
            'disable_user_tracking' => env('AZURE_MAIL_DISABLE_TRACKING', false),
            'http'                  => []
        ],
    ]

  ?>
```

Add entry to [root-of-laravel]/.env:
  
```text 
  
  #...other entries

  # Mail service entries... 
  MAIL_MAILER=azure
  
  # Azure Service entries
  AZURE_MAIL_RESOURCE_NAME=my-acs-resource-name
  # AZURE_MAIL_ENDPOINT= #optional
  AZURE_MAIL_KEY=Base64AzureAccessToken
  # AZURE_MAIL_API_VERSION=2023-03-31 #optional
  # AZURE_MAIL_DISABLE_TRACKING=false #optional
  
```


and just sent your notification mail messages!


## 📚 Documentation 

Build powerful, cloud-based communication and customer engagement experiences by adding voice, video, chat, sms, email, teams interoperability, call automation, and telephony to your apps.

Visit our Dev Site for further information regarding:
 - Azure Communication Service Docs: [English](https://learn.microsoft.com/en-us/azure/communication-services/)

 
## 💡 Last change

** [0.3.0](https://github.com/hafael/azure-mailer-driver/blob/main/CHANGELOG.md)
  * Change to Symfony Azure Bridge
  * New entry in config and .env file: "resource_name" and "AZURE_MAIL_RESOURCE_NAME"

## 📜 License 

MIT license. Copyright (c) 2023 - [Rafael](https://github.com/hafael)
For more information, see the [LICENSE](https://github.com/hafael/azure-mailer-driver/blob/main/LICENSE) file.

CHANGELOG.md
CHANGELOG
=========

0.3.0
---

 * Change to Symfony Azure Bridge
 * New entry in config and .env file: "resource_name" and "AZURE_MAIL_RESOURCE_NAME"


test/AzureMailerApiTransportTest.php
<?php

/*
 * This file is part of the Symfony package.
 *
 * (c) Fabien Potencier <fabien@symfony.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace Hafael\Azure\Transport;

use Hafael\Azure\Transport\AzureMailerApiTransport;
use PHPUnit\Framework\TestCase;
use Symfony\Component\Mailer\Envelope;
use Symfony\Component\Mailer\Header\MetadataHeader;
use Symfony\Component\Mailer\Header\TagHeader;
use Symfony\Component\Mime\Address;
use Symfony\Component\Mime\Email;
use Symfony\Component\Mime\Part\DataPart;
use Symfony\Contracts\HttpClient\HttpClientInterface;
use Symfony\Contracts\HttpClient\ResponseInterface;

class AzureMailerApiTransportTest extends TestCase
{
    /**
     * @dataProvider getTransportData
     */
    public function testToString(AzureMailerApiTransport $transport, string $expected)
    {
        $this->assertSame($expected, (string) $transport);
    }

    public static function getTransportData()
    {
        return [
            [
                new AzureMailerApiTransport('ENDPOINT', 'KEY', 'API_VERSION', 'ENGAGEMENT_TRACKING'),
                'azure+api://my-acs-resource.communication.azure.com',
            ],
            [
                (new AzureMailerApiTransport('ENDPOINT', 'KEY', 'API_VERSION', 'ENGAGEMENT_TRACKING'))->setHost('my-acs-resource.communication.azure.com'),
                'azure+api://my-acs-resource.communication.azure.com',
            ],
            [
                (new AzureMailerApiTransport('ENDPOINT', 'KEY', 'API_VERSION', 'ENGAGEMENT_TRACKING'))->setHost('my-acs-resource.communication.azure.com')->setPort(99),
                'azure+api://my-acs-resource.communication.azure.com:99',
            ],
        ];
    }

    public function testSend()
    {
        $email = new Email();
        $email->from(new Address('foo@example.com', 'Ms. Foo Bar'))
            ->to(new Address('bar@example.com', 'Mr. Recipient'))
            ->bcc('baz@example.com')
            ->text('content');

        $response = $this->createMock(ResponseInterface::class);

        $response
            ->expects($this->once())
            ->method('getStatusCode')
            ->willReturn(202);
        $response
            ->expects($this->once())
            ->method('getHeaders')
            ->willReturn(['x-message-id' => '1']);

        $httpClient = $this->createMock(HttpClientInterface::class);

        $httpClient
            ->expects($this->once())
            ->method('request')
            ->with('POST', 'https://my-acs-resource.communication.azure.com/emails:send?api-version=2023-03-31', [
                'body' => [
                    'personalizations' => [
                        [
                            'to' => [[
                                'email' => 'bar@example.com',
                                'name' => 'Mr. Recipient',
                            ]],
                            'subject' => null,
                            'bcc' => [['email' => 'baz@example.com']],
                        ],
                    ],
                    'from' => [
                        'email' => 'foo@example.com',
                        'name' => 'Ms. Foo Bar',
                    ],
                    'content' => [
                        ['type' => 'text/plain', 'value' => 'content'],
                    ],
                ],
                'headers' => [
                    'Authorization' => 'foo'
                ],
            ])
            ->willReturn($response);

        $mailer = new AzureMailerApiTransport('foo', $httpClient);
        $mailer->send($email);
    }

    public function testLineBreaksInEncodedAttachment()
    {
        $email = new Email();
        $email->from('foo@example.com')
            ->to('bar@example.com')
            // even if content doesn't include new lines, the base64 encoding performed later may add them
            ->addPart(new DataPart('Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod', 'lorem.txt'));

        $response = $this->createMock(ResponseInterface::class);

        $response
            ->expects($this->once())
            ->method('getStatusCode')
            ->willReturn(202);
        $response
            ->expects($this->once())
            ->method('getHeaders')
            ->willReturn(['operation-id' => '1']);

        $httpClient = $this->createMock(HttpClientInterface::class);

        $httpClient
            ->expects($this->once())
            ->method('request')
            ->with('POST', 'https://my-acs-resource.communication.azure.com/emails:send?api-version=2023-03-31', [
                'body' => [
                    'personalizations' => [
                        [
                            'to' => [['email' => 'bar@example.com']],
                            'subject' => null,
                        ],
                    ],
                    'from' => ['email' => 'foo@example.com'],
                    'content' => [],
                    'attachments' => [
                        [
                            'content' => 'TG9yZW0gaXBzdW0gZG9sb3Igc2l0IGFtZXQsIGNvbnNlY3RldHVyIGFkaXBpc2NpbmcgZWxpdCwgc2VkIGRvIGVpdXNtb2Q=',
                            'filename' => 'lorem.txt',
                            'type' => 'application/octet-stream',
                            'disposition' => 'attachment',
                        ],
                    ],
                ],
                'headers' => [
                    'Authorization' => 'foo'
                ],
            ])
            ->willReturn($response);

        $mailer = new AzureMailerApiTransport('ENDPOINT', 'KEY', 'API_VERSION', 'ENGAGEMENT_TRACKING', $httpClient);

        $mailer->send($email);
    }

    public function testCustomHeader()
    {
        $email = new Email();
        $email->getHeaders()->addTextHeader('foo', 'bar');
        $envelope = new Envelope(new Address('alice@system.com'), [new Address('bob@system.com')]);

        $transport = new AzureMailerApiTransport('ENDPOINT', 'KEY', 'API_VERSION', 'ENGAGEMENT_TRACKING');
        $method = new \ReflectionMethod(AzureMailerApiTransport::class, 'getPayload');
        $payload = $method->invoke($transport, $email, $envelope);

        $this->assertArrayHasKey('headers', $payload);
        $this->assertArrayHasKey('foo', $payload['headers']);
        $this->assertEquals('bar', $payload['headers']['foo']);
    }

    public function testReplyTo()
    {
        $from = 'from@example.com';
        $to = 'to@example.com';
        $replyTo = 'replyto@example.com';
        $email = new Email();
        $email->from($from)
            ->to($to)
            ->replyTo($replyTo)
            ->text('content');
        $envelope = new Envelope(new Address($from), [new Address($to)]);

        $transport = new AzureMailerApiTransport('ENDPOINT', 'KEY', 'API_VERSION', 'ENGAGEMENT_TRACKING');
        $method = new \ReflectionMethod(AzureMailerApiTransport::class, 'getPayload');
        $payload = $method->invoke($transport, $email, $envelope);

        $this->assertArrayHasKey('from', $payload);
        $this->assertArrayHasKey('email', $payload['from']);
        $this->assertSame($from, $payload['from']['email']);

        $this->assertArrayHasKey('reply_to', $payload);
        $this->assertArrayHasKey('email', $payload['reply_to']);
        $this->assertSame($replyTo, $payload['reply_to']['email']);
    }

    public function testEnvelopeSenderAndRecipients()
    {
        $from = 'from@example.com';
        $to = 'to@example.com';
        $envelopeFrom = 'envelopefrom@example.com';
        $envelopeTo = 'envelopeto@example.com';
        $email = new Email();
        $email->from($from)
            ->to($to)
            ->cc('cc@example.com')
            ->bcc('bcc@example.com')
            ->text('content');
        $envelope = new Envelope(new Address($envelopeFrom), [new Address($envelopeTo)]);

        $transport = new AzureMailerApiTransport('ENDPOINT', 'KEY', 'API_VERSION', 'ENGAGEMENT_TRACKING');
        $method = new \ReflectionMethod(AzureMailerApiTransport::class, 'getPayload');
        $payload = $method->invoke($transport, $email, $envelope);

        $this->assertArrayHasKey('from', $payload);
        $this->assertArrayHasKey('email', $payload['from']);
        $this->assertSame($envelopeFrom, $payload['from']['email']);

        $this->assertArrayHasKey('personalizations', $payload);
        $this->assertArrayHasKey('to', $payload['personalizations'][0]);
        $this->assertArrayHasKey('email', $payload['personalizations'][0]['to'][0]);
        $this->assertCount(1, $payload['personalizations'][0]['to']);
        $this->assertSame($envelopeTo, $payload['personalizations'][0]['to'][0]['email']);
    }

    public function testTagAndMetadataHeaders()
    {
        $email = new Email();
        $email->getHeaders()->add(new TagHeader('category-one'));
        $email->getHeaders()->add(new MetadataHeader('Color', 'blue'));
        $email->getHeaders()->add(new MetadataHeader('Client-ID', '12345'));
        $envelope = new Envelope(new Address('alice@system.com'), [new Address('bob@system.com')]);

        $transport = new AzureMailerApiTransport('ENDPOINT', 'KEY', 'API_VERSION', 'ENGAGEMENT_TRACKING');
        $method = new \ReflectionMethod(AzureMailerApiTransport::class, 'getPayload');
        $payload = $method->invoke($transport, $email, $envelope);

        $this->assertArrayHasKey('categories', $payload);
        $this->assertArrayHasKey('custom_args', $payload['personalizations'][0]);

        $this->assertCount(1, $payload['categories']);
        $this->assertCount(2, $payload['personalizations'][0]['custom_args']);

        $this->assertSame(['category-one'], $payload['categories']);
        $this->assertSame('blue', $payload['personalizations'][0]['custom_args']['Color']);
        $this->assertSame('12345', $payload['personalizations'][0]['custom_args']['Client-ID']);
    }
}

test/AzureMailerTransportFactoryTest.php
<?php

/*
 * This file is part of the Symfony package.
 *
 * (c) Fabien Potencier <fabien@symfony.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

 namespace Hafael\Azure\Transport;

use Hafael\Azure\Transport\AzureMailerApiTransport;
use Hafael\Azure\Transport\AzureMailerTransportFactory;
use Psr\Log\NullLogger;
use Symfony\Component\HttpClient\MockHttpClient;
use Symfony\Component\Mailer\Test\TransportFactoryTestCase;
use Symfony\Component\Mailer\Transport\Dsn;
use Symfony\Component\Mailer\Transport\TransportFactoryInterface;

class AzureMailerTransportFactoryTest extends TransportFactoryTestCase
{
    public function getFactory(): TransportFactoryInterface
    {
        return new AzureMailerTransportFactory(null, new MockHttpClient(), new NullLogger());
    }

    public static function supportsProvider(): iterable
    {
        yield [
            new Dsn('azure+api', 'default'),
            true,
        ];

        yield [
            new Dsn('azure', 'default'),
            true,
        ];
    }

    public static function createProvider(): iterable
    {
        $logger = new NullLogger();

        yield [
            new Dsn('azure+api', 'default', self::USER),
            new AzureMailerApiTransport(self::USER, new MockHttpClient(), null, $logger),
        ];

        yield [
            new Dsn('azure+api', 'example.com', self::USER, '', 8080),
            (new AzureMailerApiTransport(self::USER, new MockHttpClient(), null, $logger))->setHost('example.com')->setPort(8080),
        ];

        yield [
            new Dsn('azure', 'default', self::USER),
            new AzureMailerApiTransport(self::USER, null, $logger),
        ];

    }

    public static function unsupportedSchemeProvider(): iterable
    {
        yield [
            new Dsn('azure+foo', 'azure', self::USER),
            'The "azure+foo" scheme is not supported; supported schemes for mailer "azure" are: "azure", "azure+api".',
        ];
    }

    public static function incompleteDsnProvider(): iterable
    {
        yield [new Dsn('azure+api', 'default')];
    }
}

src/AzureMailerServiceProvider.php
<?php

namespace Hafael\Mailer\Azure;

use Illuminate\Mail\MailServiceProvider;
use Symfony\Component\Mailer\Bridge\Azure\Transport\AzureTransportFactory;
use Symfony\Component\Mailer\Transport\Dsn;

class AzureMailerServiceProvider extends MailServiceProvider
{
    /**
     * Register the Illuminate mailer instance.
     *
     * @return void
     */
    protected function registerIlluminateMailer()
    {
        $this->app->singleton('mail.manager', function($app) {
            return new AzureMailerManager($app);
        });

        $this->app->bind('mailer', function ($app) {
            return $app->make('mail.manager')->mailer();
        });

        $this->app->extend('azure', function ($app) {
            return $app->make('mail.manager')->mailer();
        });

        $this->app->extend('azure', function () {
            return (new AzureTransportFactory)->create(
                new Dsn(
                    'azure+api',
                    'default',
                    config('mail.mailers.azure.resource_name'),
                    config('mail.mailers.azure.access_key'),
                    null,
                    [
                        'api_version' => config('mail.mailers.azure.api_version'),
                        'disable_tracking' => config('mail.mailers.azure.disable_user_tracking')
                    ]
                )
            );
        });
    }
}


src/AzureMailerTransportFactory.php
<?php

namespace Hafael\Azure\Transport;

use Symfony\Component\Mailer\Exception\UnsupportedSchemeException;
use Symfony\Component\Mailer\Transport\AbstractTransportFactory;
use Symfony\Component\Mailer\Transport\Dsn;
use Symfony\Component\Mailer\Transport\TransportInterface;

final class AzureMailerTransportFactory extends AbstractTransportFactory
{
    public function create(Dsn $dsn): TransportInterface
    {
        $scheme = $dsn->getScheme();
        $endpoint = $dsn->getHost();
        $key = $this->getPassword($dsn);
        $apiVersion = $this->getApiVersion($dsn);
        $engagementTracking = $this->getEngagementTracking($dsn);

        if ('azure+api' === $scheme || 'azure' === $scheme) {
            return new AzureMailerApiTransport($endpoint, $key, $apiVersion, $engagementTracking, $this->client, $this->dispatcher, $this->logger);
        }

        throw new UnsupportedSchemeException($dsn, 'azure', $this->getSupportedSchemes());
    }

    protected function getSupportedSchemes(): array
    {
        return ['azure', 'azure+api'];
    }

    protected function getApiVersion(Dsn $dsn): string
    {
        return $dsn->getOption('api-version', '2023-03-31');
    }

    protected function getEngagementTracking(Dsn $dsn): bool
    {
        return boolval($dsn->getOption('tracking', false));
    }
}

src/AzureMailerApiTransport.php
<?php

namespace Hafael\Azure\Transport;

use Psr\EventDispatcher\EventDispatcherInterface;
use Psr\Log\LoggerInterface;
use Symfony\Component\Mailer\Envelope;
use Symfony\Component\Mailer\Exception\HttpTransportException;
use Symfony\Component\Mailer\SentMessage;
use Symfony\Component\Mailer\Transport\AbstractApiTransport;
use Symfony\Component\Mime\Address;
use Symfony\Component\Mime\Email;
use Symfony\Contracts\HttpClient\Exception\DecodingExceptionInterface;
use Symfony\Contracts\HttpClient\Exception\TransportExceptionInterface;
use Symfony\Contracts\HttpClient\HttpClientInterface;
use Symfony\Contracts\HttpClient\ResponseInterface;

class AzureMailerApiTransport extends AbstractApiTransport
{
    private const HOST = 'my-acs-resource.communication.azure.com';

    /**
     * Guzzle client instance.
     *
     * @var HttpClientInterface
     */
    protected $client;

    /**
     * User Access Key from Azure Communication Service (Primary or Secondary key).
     *
     * @var string
     */
    private string $key;

    /**
     * The endpoint API URL to which to POST emails to Azure 
     * https://{communicationServiceName}.communication.azure.com/.
     *
     * @var string
     */
    private string $endpoint;

    /**
     * The version of API to invoke.
     *
     * @var string
     */
    private string $apiVersion = '2023-03-31';

    /**
     * Indicates whether user engagement tracking should be disabled.
     *
     * @var bool
     */
    protected $disableUserEngagementTracking = false;

    /**
     * Create a new Azure Mailer Transport instance.
     *
     * @param  string  $endpoint
     * @param  string  $key
     * @param  string|null  $apiVersion
     * @param  bool  $engagementTracking
     * @param  HttpClientInterface  $client
     * @param  EventDispatcherInterface  $dispatcher
     * @param  LoggerInterface  $logger
     * @return void
     */
    public function __construct(string $endpoint, string $key, string $apiVersion = null, bool $engagementTracking = true, HttpClientInterface $client, EventDispatcherInterface $dispatcher = null, LoggerInterface $logger = null)
    {
        $this->client = $client;
        $this->endpoint = $endpoint;
        $this->key = $key;
        $this->apiVersion = empty($apiVersion) ? self::$apiVersion : $apiVersion;
        $this->disableUserEngagementTracking = !$engagementTracking;
        parent::__construct($client, $dispatcher, $logger);
    }

    /**
     * 
     * @return string
     */
    public function __toString(): string
    {
        return sprintf('azure+api://%s', $this->getAzureCSEndpoint());
    }

    /**
     * Queues an email message to be sent to one or more recipients.
     * 
     * @return ResponseInterface
     */
    protected function doSendApi(SentMessage $sentMessage, Email $email, Envelope $envelope): ResponseInterface
    {
     
        $endpointURL = $this->getEndpointRequestURL();        
        $messagePayload = $this->getMessagePayload($email, $envelope);
        $headers = $this->getSignedHeaders($messagePayload, $email);
        
        $requestOptions = [
            'body' => json_encode($messagePayload),
            'headers' => $headers,
        ];

        $response = $this->client->request('POST', $endpointURL, $requestOptions);

        try {
            $statusCode = $response->getStatusCode();
        } catch (TransportExceptionInterface $e) {
            throw new HttpTransportException('Could not reach the remote Azure server.', $response, 0, $e);
        }

        if (202 !== $statusCode) {
            try {
                $result = $response->toArray(false);
                throw new HttpTransportException('Unable to send an email (' . $result['error']['code'] . '): ' . $result['error']['message'], $response, $statusCode);
            } catch (DecodingExceptionInterface $e) {
                throw new HttpTransportException('Unable to send an email: '.$response->getContent(false).sprintf(' (code %d).', $statusCode), $response, 0, $e);
            }
        }

        $sentMessage->setMessageId(json_decode($response->getContent(false), true)['id']);
       
        return $response;
    }

    /**
     * Get the message request body
     * 
     * @param Email $email
     * @param Envelope $envelope
     * @return array
     */
    private function getMessagePayload(Email $email, Envelope $envelope): array
    {
        $addressStringifier = function (Address $address) {
            $stringified = ['address' => $address->getAddress()];

            if ($address->getName()) {
                $stringified['displayName'] = $address->getName();
            }

            return $stringified;
        };

        $data = [
            'content' => [
                'html' => $email->getHtmlBody(),
                'plainText' => $email->getTextBody(),
                'subject' => $email->getSubject(),
            ],
            'recipients' => [
                'to' => array_map($addressStringifier, $this->getRecipients($email, $envelope)),
            ],
            'senderAddress' => $envelope->getSender()->getAddress(),
            'attachments' => $this->getMessageAttachments($email),
            'userEngagementTrackingDisabled' => $this->getDisableUserEngagementTracking(),
            'headers' => empty($headers = $this->getMessageCustomHeaders($email)) ? null : $headers,
            'importance' => $this->getPriorityLevel($email->getPriority()),
        ];

        if ($emails = array_map($addressStringifier, $email->getCc())) {
            $data['recipients']['cc'] = $emails;
        }

        if ($emails = array_map($addressStringifier, $email->getBcc())) {
            $data['recipients']['bcc'] = $emails;
        }

        if ($emails = array_map($addressStringifier, $email->getReplyTo())) {
            $data['replyTo'] = $emails;
        }

        return $data;
    }

    /**
     * List of attachments. Please note that we limit the total size of 
     * an email request (which includes attachments) to 10MB. 
     * 
     * @param Email $email
     * @return array
     */
    private function getMessageAttachments(Email $email): array
    {
        $attachments = [];
        foreach ($email->getAttachments() as $attachment) {
            $headers = $attachment->getPreparedHeaders();
            $filename = $headers->getHeaderParameter('Content-Disposition', 'filename');
            $disposition = $headers->getHeaderBody('Content-Disposition');

            $att = [
                'name'            => $filename,
                'contentInBase64' => base64_encode(str_replace("\r\n", '', $attachment->bodyToString())),
                'contentType'     => $headers->get('Content-Type')->getBody(),
            ];

            if ('inline' === $disposition) {
                $att['content_id'] = $filename;
            }

            $attachments[] = $att;
        }

        return $attachments;
    }

    /**
     * Version of ACS API to invoke.
     * 
     * @return string
     */
    private function getApiVersion(): string
    {
        return $this->apiVersion;
    }

    /**
     * The communication request path with query string, for example /emails:send?api-version=2023-03-31
     * 
     * @return string
     */
    private function getURIPathAndQuery(): string
    {
        return '/emails:send?api-version=' . $this->getApiVersion();
    }
    
    /**
     * The communication resource, for example https://my-acs-resource.communication.azure.com
     * 
     * @return string
     */
    private function getEndpointRequestURL(): string
    {
        return $this->getAzureCSEndpoint() . $this->getURIPathAndQuery();
    }
    
    /**
     * The communication domain host, for example my-acs-resource.communication.azure.com
     * 
     * @return string
     */
    private function getAzureCSEndpoint(): string
    {
        return ($this->endpoint ?: self::HOST);
    }

    /**
     * Get name and email from contact list
     * 
     * @param array $contacs
     * @return array
     */
    private function getContactNameAndEmail($contacts): array
    {
        $formatted = [];
        if (empty($contacts)) {
            return [];
        }
        foreach ($contacts as $address => $display) {
            $formatted[] =  array_filter([
                'displayName' => $display,
                'email'       => $address,
            ]);
        }
        return $formatted;
    }

    /**
     * Generate hash from message content
     * 
     * @param string $content
     * @return string
     */
    private function getContentHash(string $content): string
    {
        //$data = mb_convert_encoding($content, 'UTF-8');
        $hash = hash('sha256', $content, true);
        //return base64 hash
        return base64_encode($hash);
    }

    /**
     * Generate sha256 hash and encode to base64 to produces the digest string.
     * 
     * @return string
     */
    private function getAuthenticationSignature(string $content): string
    {
        //dd($content);
        $key = base64_decode($this->key);
        $hashedBytes = hash_hmac('sha256', mb_convert_encoding($content, 'UTF-8'), $key, true);
        return base64_encode($hashedBytes);
    }
   
    /**
     * Get authenticated headers for signed request,
     * 
     * @return array
     */
    private function getSignedHeaders(array $payload, Email $message): array
    {
        //HTTP Method verb (uppercase)
        $verb = "POST";

        //Request time
        $datetime = new \DateTime("now", new \DateTimeZone("UTC"));
        $utcNow = $datetime->format('D, d M Y H:i:s \G\M\T');
        
        //Content hash signature
        $contentHash = $this->getContentHash(json_encode($payload));
        
        //ACS Endpoint
        $host = str_replace('https://', '', $this->endpoint);
        
        //Sendmail endpoint from communication email delivery service
        $urlPathAndQuery = '/emails:send?api-version=' . $this->getApiVersion();
        
        //Signed request headers
        $stringToSign = "{$verb}\n{$urlPathAndQuery}\n{$utcNow};{$host};{$contentHash}";
        
        //Authenticate headers with ACS primary or secondary key
        $signature = $this->getAuthenticationSignature($stringToSign);
        
        //get GUID part of message id to identify the long running operation
        //$messageId = explode('@', $message->generateMessageId(), 2)[0];

        $messageId = $this->generateMessageId();

        $headers = [
            'Content-Type' => 'application/json',
            'repeatability-request-id' => $messageId,
            'Operation-Id' => $messageId,
            'repeatability-first-sent' => $utcNow,
            'x-ms-date' => $utcNow,
            'x-ms-content-sha256' => $contentHash,
            'x-ms-client-request-id' => $messageId,
            "Authorization" => "HMAC-SHA256 SignedHeaders=x-ms-date;host;x-ms-content-sha256&Signature={$signature}",
        ];

        //return signed headers to http client request
        return $headers;
    }

    /**
     * Identify the long running operation.
     * 
     * @return array
     */
    private function generateMessageId(): string
    {
        $data = random_bytes(16);
        assert(strlen($data) == 16);
        $data[6] = chr(ord($data[6]) & 0x0f | 0x40);
        $data[8] = chr(ord($data[8]) & 0x3f | 0x80);

        // Output the 36 character UUID.
        return vsprintf('%s%s-%s-%s-%s-%s%s%s', str_split(bin2hex($data), 4));
    }
    
    /**
     * Custom email headers to be passed.
     * 
     * @return array
     */
    private function getMessageCustomHeaders(Email $email): array
    {

        $headers = [];

        $headersToBypass = ['x-ms-client-request-id', 'operation-id', 'authorization', 'x-ms-content-sha256', 'received', 'dkim-signature', 'content-transfer-encoding', 'from', 'to', 'cc', 'bcc', 'subject', 'content-type', 'reply-to'];
        
        foreach ($email->getHeaders()->all() as $name => $header) {
            if (\in_array($name, $headersToBypass, true)) {
                continue;
            }
            $headers[$header->getName()] = $header->getBodyAsString();
        }

        return $headers;
    }
    
    /**
     * Get the message priority level
     * 
     * @return string
     */
    private function getPriorityLevel($priority): ?string
    {
        return strtolower([
            Email::PRIORITY_HIGHEST => 'Highest',
            Email::PRIORITY_HIGH    => 'High',
            Email::PRIORITY_NORMAL  => 'Normal',
            Email::PRIORITY_LOW     => 'Low',
            Email::PRIORITY_LOWEST  => 'Lowest',
        ][$priority]);
    }

    /**
     * Indicates whether user engagement tracking should be disabled.
     * 
     * @return bool
     */
    private function getDisableUserEngagementTracking(): bool
    {
        return $this->disableUserEngagementTracking;
    }
}